In [1]:
import pandas as pd

In [2]:
pd.read_json('Telco-Customer-Churn.json')

customerID Churn                                           customer  \
0     0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1     0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2     0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3     0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4     0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
...          ...   ...                                                ...   
7262  9987-LUTYD    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7263  9992-RRAMN   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7264  9992-UJOEL    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7265  9993-LHIEB    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7266  9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
0        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1       {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
...                                                 ...   
7262     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7263    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7264     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7265     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7266  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                               internet  \
0     {'InternetService': 'DSL', 'OnlineSecurity': '...   
1     {'InternetService': 'DSL', 'OnlineSecurity': '...   
2     {'InternetService': 'Fiber optic', 'OnlineSecu...   
3     {'InternetService': 'Fiber optic', 'OnlineSecu...   
4     {'InternetService': 'Fiber optic', 'OnlineSecu...   
...                                                 ...   
7262  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7263  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7264  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7265  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7266  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  
0     {'Contract': 'One year', 'PaperlessBilling': '...  
1     {'Contract': 'Month-to-month', 'PaperlessBilli...  
2     {'Contract': 'Month-to-month', 'PaperlessBilli...  
3     {'Contract': 'Month-to-month', 'PaperlessBilli...  
4     {'Contract': 'Month-to-month', 'PaperlessBilli...  
...                                                 ...  
7262  {'Contract': 'One year', 'PaperlessBilling': '...  
7263  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7264  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7265  {'Contract': 'Two year', 'PaperlessBilling': '...  
7266  {'Contract': 'Two year', 'PaperlessBilling': '...  

[7267 rows x 6 columns]

In [3]:
import json
with open('Telco-Customer-Churn.json') as f:
    json_bruto = json.load(f)
df = pd.json_normalize(json_bruto)
df

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                       0   
7263  9992-RRAMN   Yes            Male                       0   
7264  9992-UJOEL    No            Male                       0   
7265  9993-LHIEB    No            Male                       0   
7266  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7262               No                  No               13                Yes   
7263              Yes                  No               22                Yes   
7264               No                  No                2                Yes   
7265              Yes                 Yes               67                Yes   
7266              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
0                     No                      DSL  ...                   Yes   
1                    Yes                      DSL  ...                    No   
2                     No              Fiber optic  ...                    No   
3                     No              Fiber optic  ...                   Yes   
4                     No              Fiber optic  ...                    No   
...                  ...                      ...  ...                   ...   
7262                  No                      DSL  ...                    No   
7263                 Yes              Fiber optic  ...                    No   
7264                  No                      DSL  ...                   Yes   
7265                  No                      DSL  ...                    No   
7266    No phone service                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7262                        No                  Yes                   No   
7263                        No                   No                   No   
7264                        No                   No                   No   
7265                       Yes                  Yes                   No   
7266                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                    

## Informações dos dados

Os dados trabalham sobre, principalmente, o Churn dos clientes da empresa, detalhando o sexo, informações sobre serem senis, tempo de fidelidade, e outros aspectos como os serviços que foram prestados a eles, tipo de contrato, pagamento, preço, etc. Os dados serão aplicados em ML para conseguir aprender sob quais circunstâcias o churn costuma ter mais frequência, a fim de diminuir o mesmo.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

### Dados Nulos e Vazios:

In [5]:
df.isnull().sum().sum()

0

In [6]:
for col in df.columns:
    if (df[col] == '').any():
        print(f'Coluna: {col}: ')
    else:
        print(0)

0
Coluna: Churn: 
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0


In [7]:
for col in df.columns:
    if (df[col] == ' ').any():
        print(f'Coluna: {col}: ')
    else:
        print(0)

0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
Coluna: account.Charges.Total: 


In [8]:
churn_vazio = df[df['Churn'] == '']
churn_vazio[['Churn']]

Churn
30        
75        
96        
98        
175       
...    ...
7158      
7180      
7211      
7239      
7247      

[224 rows x 1 columns]

In [9]:
contas_total_vazio = df[df['account.Charges.Total'] == ' ']
contas_total_vazio[['account.Charges.Total']]

account.Charges.Total
975                       
1775                      
1955                      
2075                      
2232                      
2308                      
2930                      
3134                      
3203                      
4169                      
5599

### Outliers e Duplicatas:

In [10]:
df.describe()

customer.SeniorCitizen  customer.tenure  account.Charges.Monthly
count             7267.000000      7267.000000              7267.000000
mean                 0.162653        32.346498                64.720098
std                  0.369074        24.571773                30.129572
min                  0.000000         0.000000                18.250000
25%                  0.000000         9.000000                35.425000
50%                  0.000000        29.000000                70.300000
75%                  0.000000        55.000000                89.875000
max                  1.000000        72.000000               118.750000

In [11]:
def calcula_outliers(dataframe, coluna):
    Q1 = dataframe[coluna].quantile(.25)
    Q3 = dataframe[coluna].quantile(.75)
    IQR = Q3 - Q1
    valor_min = Q1 - IQR*1.5
    valor_max = Q3 + IQR*1.5
    
    outlier = (dataframe[coluna] < valor_min) | (dataframe[coluna] > valor_max)
    return outlier

In [12]:
outliers_index = calcula_outliers(df, 'customer.tenure')
outliers_index

0       False
1       False
2       False
3       False
4       False
        ...  
7262    False
7263    False
7264    False
7265    False
7266    False
Name: customer.tenure, Length: 7267, dtype: bool

In [13]:
df[outliers_index]['customer.tenure']

Series([], Name: customer.tenure, dtype: int64)

In [14]:
outliers_index = calcula_outliers(df, 'account.Charges.Monthly')
df[outliers_index]['account.Charges.Monthly']

Series([], Name: account.Charges.Monthly, dtype: float64)

In [15]:
df.duplicated().sum()

0

Não há outliers nas colunas conferidas, duplicatas, ou dados nulos, mas existem dados vazios nas colunas Churn e account.Charges.Total, agora para conferir se encontrarei outliers nesta última coluna, teremos que tratar dos dados vazios, como ele é o resultado de toda a receita do cliente durante a vida, basta mutiplicar os dados que equivalem ao preço por mês com o tempo total do cliente com a empresa, já os dados vazios do churn não é possível substituí-los, então terei de removê-los. Também há de se notar que não existem colunas duplicadas.

In [16]:
df_limpo = df.drop(churn_vazio.index).copy()
df_limpo

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                       0   
7263  9992-RRAMN   Yes            Male                       0   
7264  9992-UJOEL    No            Male                       0   
7265  9993-LHIEB    No            Male                       0   
7266  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7262               No                  No               13                Yes   
7263              Yes                  No               22                Yes   
7264               No                  No                2                Yes   
7265              Yes                 Yes               67                Yes   
7266              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
0                     No                      DSL  ...                   Yes   
1                    Yes                      DSL  ...                    No   
2                     No              Fiber optic  ...                    No   
3                     No              Fiber optic  ...                   Yes   
4                     No              Fiber optic  ...                    No   
...                  ...                      ...  ...                   ...   
7262                  No                      DSL  ...                    No   
7263                 Yes              Fiber optic  ...                    No   
7264                  No                      DSL  ...                   Yes   
7265                  No                      DSL  ...                    No   
7266    No phone service                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7262                        No                  Yes                   No   
7263                        No                   No                   No   
7264                        No                   No                   No   
7265                       Yes                  Yes                   No   
7266                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                    

In [17]:
df_limpo.reset_index(drop=True, inplace=True)
contas_total_vazio = df_limpo['account.Charges.Total'] == ' '
df_limpo[contas_total_vazio][
    ['account.Charges.Total', 'customer.tenure', 'account.Charges.Monthly']
    ]

account.Charges.Total  customer.tenure  account.Charges.Monthly
945                                       0                    56.05
1731                                      0                    20.00
1906                                      0                    61.90
2025                                      0                    19.70
2176                                      0                    20.25
2250                                      0                    25.35
2855                                      0                    73.35
3052                                      0                    25.75
3118                                      0                    52.55
4054                                      0                    80.85
5433                                      0                    19.85

In [18]:
ternure_nulo = df_limpo['customer.tenure'] == 0
df_limpo[ternure_nulo][
    ['account.Charges.Total', 'customer.tenure', 'account.Charges.Monthly', 'account.Contract', 'Churn']
    ]

account.Charges.Total  customer.tenure  account.Charges.Monthly  \
945                                       0                    56.05   
1731                                      0                    20.00   
1906                                      0                    61.90   
2025                                      0                    19.70   
2176                                      0                    20.25   
2250                                      0                    25.35   
2855                                      0                    73.35   
3052                                      0                    25.75   
3118                                      0                    52.55   
4054                                      0                    80.85   
5433                                      0                    19.85   

     account.Contract Churn  
945          Two year    No  
1731         Two year    No  
1906         Two year    No  
2025         One year    No  
2176         Two year    No  
2250         Two year    No  
2855         Two year    No  
3052         Two year    No  
3118         Two year    No  
4054         Two year    No  
5433         Two year    No

Aqui as colunas em que o total são vazias são as mesmas que o tempo do cliente com a empresa juntos está equivalendo a zero, mas todas também possuem o contrato de 2 anos, o que me faz chegar na ideia que posso tratar os dados de duas formas: removê-los, o que fariam os dados serm perdidos no caminho, ou tentar tratar eles baseado nos contratos de dois anos.

In [19]:
contrato_dois_anos = df_limpo['account.Contract'] == 'Two year'
df_limpo[contrato_dois_anos].describe()

customer.SeniorCitizen  customer.tenure  account.Charges.Monthly
count             1695.000000      1695.000000              1695.000000
mean                 0.085546        56.735103                60.770413
std                  0.279775        18.209363                34.678865
min                  0.000000         0.000000                18.400000
25%                  0.000000        48.000000                24.025000
50%                  0.000000        64.000000                64.350000
75%                  0.000000        71.000000                90.450000
max                  1.000000        72.000000               118.750000

In [20]:
df_aux = df_limpo.drop(df_limpo[contas_total_vazio].index).copy()
df_aux[df_aux['account.Contract'] == 'Two year'].describe()

customer.SeniorCitizen  customer.tenure  account.Charges.Monthly
count             1685.000000       1685.00000              1685.000000
mean                 0.086053         57.07181                60.872374
std                  0.280526         17.72913                34.712210
min                  0.000000          1.00000                18.400000
25%                  0.000000         48.00000                24.050000
50%                  0.000000         64.00000                64.450000
75%                  0.000000         71.00000                90.550000
max                  1.000000         72.00000               118.750000

In [21]:
df_contrato_two_year = df_aux[df_aux['account.Contract'] == 'Two year']
outliers_contrato_two_year = calcula_outliers(df_contrato_two_year, 'customer.tenure')
df_contrato_two_year[outliers_contrato_two_year]


customerID Churn customer.gender  customer.SeniorCitizen  \
38    0064-SUDOG    No          Female                       0   
132   0224-NIJLP    No            Male                       0   
208   0323-XWWTN    No            Male                       0   
309   0458-HEUZG    No          Female                       0   
328   0486-LGCCH    No            Male                       0   
...          ...   ...             ...                     ...   
6467  9152-AMKAK    No            Male                       0   
6575  9350-ZXYJC    No          Female                       0   
6631  9424-CMPOG    No            Male                       0   
6727  9553-DLCLU    No          Female                       0   
6996  9928-BZVLZ    No          Female                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
38                Yes                 Yes               12                Yes   
132               Yes                 Yes                8                Yes   
208                No                 Yes                3                Yes   
309                No                  No               13                 No   
328               Yes                 Yes               11                Yes   
...               ...                 ...              ...                ...   
6467               No                  No                6                Yes   
6575               No                  No                4                Yes   
6631              Yes                 Yes                9                Yes   
6727               No                 Yes               13                Yes   
6996               No                  No               12                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
38                    No                       No  ...   No internet service   
132                   No                       No  ...   No internet service   
208                  Yes                       No  ...   No internet service   
309     No phone service                      DSL  ...                    No   
328                   No                       No  ...   No internet service   
...                  ...                      ...  ...                   ...   
6467                  No                      DSL  ...                    No   
6575                 Yes                       No  ...   No internet service   
6631                  No                       No  ...   No internet service   
6727                 Yes                      DSL  ...                   Yes   
6996    No phone service                      DSL  ...                    No   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
38         No internet service  No internet service  No internet service   
132        No internet service  No internet service  No internet service   
208        No internet service  No internet service  No internet service   
309                        Yes                  Yes                   No   
328        No internet service  No internet service  No internet service   
...                        ...                  ...                  ...   
6467                       Yes                  Yes                  Yes   
6575       No internet service  No internet service  No internet service   
6631       No internet service  No internet service  No internet service   
6727                       Yes                  Yes                  Yes   
6996                       Yes                  Yes                   No   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
38        No internet service         Two year                       No   
132       No internet service         Two year                       No   
208       No internet service         Two year                      Yes   
309                        No         Two year                    

Considerando que todos os casos não há Churn, o desvio-padrão consideravelmente alto, porém um max que não foge, já um min que foge 69x dele, vemos que ele tem uma margem de 4,1% de outliers, ao retirá-los:

In [22]:
df_aux_sem_out = df_contrato_two_year.drop(df_contrato_two_year[outliers_contrato_two_year].index).copy()
filtro_contrato_churn = (df_aux_sem_out['account.Contract'] == 'Two year') & (df_aux_sem_out['Churn'] == 'No')
df_filtrado = df_aux_sem_out.loc[filtro_contrato_churn]
df_filtrado.describe()

customer.SeniorCitizen  customer.tenure  account.Charges.Monthly
count             1568.000000      1568.000000              1568.000000
mean                 0.088648        59.073980                61.483450
std                  0.284326        15.029251                34.520375
min                  0.000000        14.000000                18.400000
25%                  0.000000        51.000000                24.250000
50%                  0.000000        65.000000                65.175000
75%                  0.000000        71.000000                90.462500
max                  1.000000        72.000000               118.750000

Percebe-se que também a quantia de contratos de dois anos que não possui o Churn com e sem os outliers é dominante, sendo de cerca de 97%. Também pode-se notar que houve um certo "boom" no período entre 3 e 2 anos antes da coleta dos dados no contrato de dois anos (3 anos antes da coleta sendo o início de quando os dados começaram a entrar), e houve uma repentina queda da assinatura desse contrato. Pela ínfima quantidade de dados vazios que estou trabalhando (11 no total), e apesar da alta chance de eles serem desse período de 2/3 anos antes da coleta dos dados, não há nada que possa confirmar isso, eu optei por tirá-los do dataframe, e irei trabalhar sem eles, porém com os outliers, que apesar de fugirem muito do desvio-padrão, dado ao contexto, são dados válidos.

In [23]:
df_aux['account.Charges.Total'] = df_aux['account.Charges.Total'].str.replace(r'[^0-9]', '', regex=True)

In [24]:
import numpy as np
df_aux['account.Charges.Total'] = df_aux['account.Charges.Total'].astype(np.float64)
df_aux.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7032 non-null   object 
 1   Churn                      7032 non-null   object 
 2   customer.gender            7032 non-null   object 
 3   customer.SeniorCitizen     7032 non-null   int64  
 4   customer.Partner           7032 non-null   object 
 5   customer.Dependents        7032 non-null   object 
 6   customer.tenure            7032 non-null   int64  
 7   phone.PhoneService         7032 non-null   object 
 8   phone.MultipleLines        7032 non-null   object 
 9   internet.InternetService   7032 non-null   object 
 10  internet.OnlineSecurity    7032 non-null   object 
 11  internet.OnlineBackup      7032 non-null   object 
 12  internet.DeviceProtection  7032 non-null   object 
 13  internet.TechSupport       7032 non-null   object 
 1

In [25]:
df_aux.describe()

customer.SeniorCitizen  customer.tenure  account.Charges.Monthly  \
count             7032.000000      7032.000000              7032.000000   
mean                 0.162400        32.421786                64.798208   
std                  0.368844        24.545260                30.085974   
min                  0.000000         1.000000                18.250000   
25%                  0.000000         9.000000                35.587500   
50%                  0.000000        29.000000                70.350000   
75%                  0.000000        55.000000                89.862500   
max                  1.000000        72.000000               118.750000   

       account.Charges.Total  
count            7032.000000  
mean           126251.908419  
std            192523.026544  
min                19.000000  
25%              7562.000000  
50%             36769.000000  
75%            143010.000000  
max            867245.000000

In [26]:
outliers_total = calcula_outliers(df_aux, 'account.Charges.Total')
df_aux[outliers_total]

customerID Churn customer.gender  customer.SeniorCitizen  \
6     0013-SMEOE    No          Female                       1   
11    0017-IUDMW    No          Female                       0   
13    0019-EFAEP    No          Female                       0   
26    0036-IHMOT    No          Female                       0   
32    0052-DCKON    No            Male                       0   
...          ...   ...             ...                     ...   
7019  9959-WOFKT    No            Male                       0   
7025  9968-FFVVH    No            Male                       0   
7030  9972-VAFJJ    No          Female                       1   
7034  9978-HYCIN    No            Male                       1   
7041  9993-LHIEB    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
6                 Yes                  No               71                Yes   
11                Yes                 Yes               72                Yes   
13                 No                  No               72                Yes   
26                Yes                 Yes               55                Yes   
32                Yes                  No               66                Yes   
...               ...                 ...              ...                ...   
7019               No                 Yes               71                Yes   
7025               No                  No               63                Yes   
7030              Yes                  No               53                Yes   
7034              Yes                 Yes               47                Yes   
7041              Yes                 Yes               67                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
6                     No              Fiber optic  ...                   Yes   
11                   Yes              Fiber optic  ...                   Yes   
13                   Yes              Fiber optic  ...                   Yes   
26                    No              Fiber optic  ...                   Yes   
32                   Yes              Fiber optic  ...                   Yes   
...                  ...                      ...  ...                   ...   
7019                 Yes              Fiber optic  ...                    No   
7025                 Yes                      DSL  ...                   Yes   
7030                  No              Fiber optic  ...                   Yes   
7034                  No              Fiber optic  ...                   Yes   
7041                  No                      DSL  ...                    No   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
6                          Yes                  Yes                  Yes   
11                         Yes                  Yes                  Yes   
13                         Yes                   No                  Yes   
26                         Yes                  Yes                  Yes   
32                         Yes                  Yes                  Yes   
...                        ...                  ...                  ...   
7019                       Yes                   No                  Yes   
7025                       Yes                  Yes                   No   
7030                       Yes                  Yes                  Yes   
7034                        No                   No                  Yes   
7041                       Yes                  Yes                   No   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
6                         Yes         Two year                      Yes   
11                        Yes         Two year                      Yes   
13                         No         Two year                      Yes   
26                        Yes         One year                    

In [27]:
df_aux.loc[outliers_total, 'account.Charges.Total'] = df_aux.loc[outliers_total, 'account.Charges.Monthly'] * df_aux.loc[outliers_total, 'customer.tenure']
df_aux[outliers_total][
    ['account.Charges.Total', 'account.Charges.Monthly', 'customer.tenure']
]

account.Charges.Total  account.Charges.Monthly  customer.tenure
6                   7788.70                   109.70               71
11                  8409.60                   116.80               72
13                  7293.60                   101.30               72
26                  5703.50                   103.70               55
32                  7642.80                   115.80               66
...                     ...                      ...              ...
7019                7575.70                   106.70               71
7025                4334.40                    68.80               63
7030                4982.00                    94.00               53
7034                3992.65                    84.95               47
7041                4545.95                    67.85               67

[977 rows x 3 columns]

In [28]:
df_aux[outliers_total].describe()

customer.SeniorCitizen  customer.tenure  account.Charges.Monthly  \
count              977.000000       977.000000               977.000000   
mean                 0.230297        60.076766                92.502712   
std                  0.421238        10.444587                15.330243   
min                  0.000000        32.000000                49.200000   
25%                  0.000000        52.000000                81.250000   
50%                  0.000000        62.000000                94.800000   
75%                  0.000000        70.000000               104.900000   
max                  1.000000        72.000000               118.750000   

       account.Charges.Total  
count             977.000000  
mean             5538.743142  
std              1323.788440  
min              3340.750000  
25%              4399.500000  
50%              5428.800000  
75%              6512.400000  
max              8550.000000

In [29]:
outliers_total = calcula_outliers(df_aux, 'account.Charges.Total')
df_aux[outliers_total]

customerID Churn customer.gender  customer.SeniorCitizen  \
10    0017-DINOC    No            Male                       0   
22    0027-KWYKW    No          Female                       0   
30    0048-LUMLS    No            Male                       0   
41    0068-FIGTF    No          Female                       0   
56    0096-FCPUF    No            Male                       0   
...          ...   ...             ...                     ...   
6990  9921-EZKBY    No            Male                       0   
6991  9921-QFQUL    No          Female                       0   
6995  9927-DSWDF    No            Male                       0   
6999  9931-KGHOA    No          Female                       0   
7013  9948-YPTDG   Yes            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
10                 No                  No               54                 No   
22                Yes                 Yes               23                Yes   
30                Yes                 Yes               37                Yes   
41                 No                  No               27                Yes   
56                 No                  No               30                Yes   
...               ...                 ...              ...                ...   
6990               No                 Yes               33                Yes   
6991              Yes                  No               23                Yes   
6995              Yes                  No               22                Yes   
6999              Yes                  No               46                Yes   
7013              Yes                  No               38                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
10      No phone service                      DSL  ...                    No   
22                   Yes              Fiber optic  ...                    No   
30                    No              Fiber optic  ...                    No   
41                    No                      DSL  ...                   Yes   
56                   Yes                      DSL  ...                    No   
...                  ...                      ...  ...                   ...   
6990                 Yes              Fiber optic  ...                    No   
6991                 Yes              Fiber optic  ...                    No   
6995                  No              Fiber optic  ...                   Yes   
6999                 Yes                      DSL  ...                    No   
7013                  No              Fiber optic  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
10                          No                  Yes                  Yes   
22                          No                   No                  Yes   
30                          No                   No                  Yes   
41                         Yes                  Yes                  Yes   
56                          No                   No                   No   
...                        ...                  ...                  ...   
6990                       Yes                   No                   No   
6991                       Yes                   No                   No   
6995                        No                  Yes                  Yes   
6999                        No                   No                   No   
7013                        No                   No                   No   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
10                         No         Two year                       No   
22                         No   Month-to-month                      Yes   
30                        Yes         One year                       No   
41                        Yes         One year                    

In [30]:
df_aux.loc[outliers_total, 'account.Charges.Total'] = df_aux.loc[outliers_total, 'account.Charges.Monthly'] * df_aux.loc[outliers_total, 'customer.tenure']
df_aux[outliers_total][
    ['account.Charges.Total', 'account.Charges.Monthly', 'customer.tenure']
]

account.Charges.Total  account.Charges.Monthly  customer.tenure
10                  2440.80                    45.20               54
22                  1926.25                    83.75               23
30                  3374.40                    91.20               37
41                  2111.40                    78.20               27
56                  1935.00                    64.50               30
...                     ...                      ...              ...
6990                3144.90                    95.30               33
6991                2073.45                    90.15               23
6995                2301.20                   104.60               22
6999                2530.00                    55.00               46
7013                3019.10                    79.45               38

[760 rows x 3 columns]

In [31]:
df_aux.describe()

customer.SeniorCitizen  customer.tenure  account.Charges.Monthly  \
count             7032.000000      7032.000000              7032.000000   
mean                 0.162400        32.421786                64.798208   
std                  0.368844        24.545260                30.085974   
min                  0.000000         1.000000                18.250000   
25%                  0.000000         9.000000                35.587500   
50%                  0.000000        29.000000                70.350000   
75%                  0.000000        55.000000                89.862500   
max                  1.000000        72.000000               118.750000   

       account.Charges.Total  
count            7032.000000  
mean            25172.697540  
std             32958.315374  
min                19.000000  
25%              3247.200000  
50%              7979.800000  
75%             37526.750000  
max            146065.000000

In [32]:
df_aux['account.Charges.Total'] = df_aux['account.Charges.Monthly'] * df_aux['customer.tenure']
df_aux.describe()

customer.SeniorCitizen  customer.tenure  account.Charges.Monthly  \
count             7032.000000      7032.000000              7032.000000   
mean                 0.162400        32.421786                64.798208   
std                  0.368844        24.545260                30.085974   
min                  0.000000         1.000000                18.250000   
25%                  0.000000         9.000000                35.587500   
50%                  0.000000        29.000000                70.350000   
75%                  0.000000        55.000000                89.862500   
max                  1.000000        72.000000               118.750000   

       account.Charges.Total  
count            7032.000000  
mean             2283.147248  
std              2264.703327  
min                18.800000  
25%               397.800000  
50%              1394.575000  
75%              3791.250000  
max              8550.000000

Aqui eu realizei um teste que se mostrou funcionando, eu tentei outras maneiras de ficar repetindo a função de chamar apenas os outliers, mas acontece que em algum momento estagnava fora do limite possível, que é o de 8550, pois é a multiplicação do limite max de 'customer.tenure' com 'account.Charges.Monthly', e após analisar e ver que eles são dados confiáveis, optei por resetar toda a coluna do 'account.Charges.Total' que possuia muitos dados incoerentes.

### Dados Categóricos:

In [33]:
df_aux

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7038  9987-LUTYD    No          Female                       0   
7039  9992-RRAMN   Yes            Male                       0   
7040  9992-UJOEL    No            Male                       0   
7041  9993-LHIEB    No            Male                       0   
7042  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7038               No                  No               13                Yes   
7039              Yes                  No               22                Yes   
7040               No                  No                2                Yes   
7041              Yes                 Yes               67                Yes   
7042              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
0                     No                      DSL  ...                   Yes   
1                    Yes                      DSL  ...                    No   
2                     No              Fiber optic  ...                    No   
3                     No              Fiber optic  ...                   Yes   
4                     No              Fiber optic  ...                    No   
...                  ...                      ...  ...                   ...   
7038                  No                      DSL  ...                    No   
7039                 Yes              Fiber optic  ...                    No   
7040                  No                      DSL  ...                   Yes   
7041                  No                      DSL  ...                    No   
7042    No phone service                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7038                        No                  Yes                   No   
7039                        No                   No                   No   
7040                        No                   No                   No   
7041                       Yes                  Yes                   No   
7042                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                    

In [34]:
variaveis_categoricas_binarias = []
for col in df_aux.columns:
    print(f'Coluna {col}: {df_aux[col].unique()}')
    if (len(df_aux[col].unique()) == 2):
        variaveis_categoricas_binarias.append(col)
print(variaveis_categoricas_binarias)

Coluna customerID: ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
Coluna Churn: ['No' 'Yes']
Coluna customer.gender: ['Female' 'Male']
Coluna customer.SeniorCitizen: [0 1]
Coluna customer.Partner: ['Yes' 'No']
Coluna customer.Dependents: ['Yes' 'No']
Coluna customer.tenure: [ 9  4 13  3 71 63  7 65 54 72  5 56 34  1 45 50 23 55 26 69 37 49 66 67
 20 43 59 12 27  2 25 29 14 35 64 39 40 11  6 30 70 57 58 16 32 33 10 21
 61 15 44 22 24 19 47 62 46 52  8 60 48 28 41 53 68 31 36 17 18 51 38 42]
Coluna phone.PhoneService: ['Yes' 'No']
Coluna phone.MultipleLines: ['No' 'Yes' 'No phone service']
Coluna internet.InternetService: ['DSL' 'Fiber optic' 'No']
Coluna internet.OnlineSecurity: ['No' 'Yes' 'No internet service']
Coluna internet.OnlineBackup: ['Yes' 'No' 'No internet service']
Coluna internet.DeviceProtection: ['No' 'Yes' 'No internet service']
Coluna internet.TechSupport: ['Yes' 'No' 'No internet service']
Coluna internet.StreamingTV: ['Yes' 'No' '

In [35]:
variaveis_categoricas_binarias.remove('customer.SeniorCitizen')
mapeamento = {'Yes': 1, 'No': 0, 'Male': 0, 'Female': 1}
df_aux[variaveis_categoricas_binarias] = df_aux[variaveis_categoricas_binarias].replace(mapeamento)
df_aux

customerID  Churn  customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO      0                1                       0   
1     0003-MKNFE      0                0                       0   
2     0004-TLHLJ      1                0                       0   
3     0011-IGKFF      1                0                       1   
4     0013-EXCHZ      1                1                       1   
...          ...    ...              ...                     ...   
7038  9987-LUTYD      0                1                       0   
7039  9992-RRAMN      1                0                       0   
7040  9992-UJOEL      0                0                       0   
7041  9993-LHIEB      0                0                       0   
7042  9995-HOTOH      0                0                       0   

      customer.Partner  customer.Dependents  customer.tenure  \
0                    1                    1                9   
1                    0                    0                9   
2                    0                    0                4   
3                    1                    0               13   
4                    1                    0                3   
...                ...                  ...              ...   
7038                 0                    0               13   
7039                 1                    0               22   
7040                 0                    0                2   
7041                 1                    1               67   
7042                 1                    1               63   

      phone.PhoneService phone.MultipleLines internet.InternetService  ...  \
0                      1                  No                      DSL  ...   
1                      1                 Yes                      DSL  ...   
2                      1                  No              Fiber optic  ...   
3                      1                  No              Fiber optic  ...   
4                      1                  No              Fiber optic  ...   
...                  ...                 ...                      ...  ...   
7038                   1                  No                      DSL  ...   
7039                   1                 Yes              Fiber optic  ...   
7040                   1                  No                      DSL  ...   
7041                   1                  No                      DSL  ...   
7042                   0    No phone service                      DSL  ...   

     internet.OnlineBackup internet.DeviceProtection internet.TechSupport  \
0                      Yes                        No                  Yes   
1                       No                        No                   No   
2                       No                       Yes                   No   
3                      Yes                       Yes                   No   
4                       No                        No                  Yes   
...                    ...                       ...                  ...   
7038                    No                        No                  Yes   
7039                    No                        No                   No   
7040                   Yes                        No                   No   
7041                    No                       Yes                  Yes   
7042                   Yes                       Yes                   No   

     internet.StreamingTV internet.StreamingMovies account.Contract  \
0                     Yes                       No         One year   
1                      No                      Yes   Month-to-month   
2                      No                       No   Month-to-month   
3                     Yes                      Yes   Month-to-month   
4                     Yes                       No   Month-to-month   
...                   ...                      ...              ...   
7038                   No                       No         On

Transformadas em 0 e 1 os dados categóricos binários, agora trabalharei com os nominais, já que não existem ordinais nesse dataframe, começando pelo customerID que não possui nenhuma função para o aprendizado da máquina já que se trata apenas de uma identificação de cada usuário, após isso irei usar o método get_dummies para transformar as colunas restantes em binárias.

In [36]:
df_aux.drop(columns=['customerID'], inplace=True)
df_aux

Churn  customer.gender  customer.SeniorCitizen  customer.Partner  \
0         0                1                       0                 1   
1         0                0                       0                 0   
2         1                0                       0                 0   
3         1                0                       1                 1   
4         1                1                       1                 1   
...     ...              ...                     ...               ...   
7038      0                1                       0                 0   
7039      1                0                       0                 1   
7040      0                0                       0                 0   
7041      0                0                       0                 1   
7042      0                0                       0                 1   

      customer.Dependents  customer.tenure  phone.PhoneService  \
0                       1                9                   1   
1                       0                9                   1   
2                       0                4                   1   
3                       0               13                   1   
4                       0                3                   1   
...                   ...              ...                 ...   
7038                    0               13                   1   
7039                    0               22                   1   
7040                    0                2                   1   
7041                    1               67                   1   
7042                    1               63                   0   

     phone.MultipleLines internet.InternetService internet.OnlineSecurity  \
0                     No                      DSL                      No   
1                    Yes                      DSL                      No   
2                     No              Fiber optic                      No   
3                     No              Fiber optic                      No   
4                     No              Fiber optic                      No   
...                  ...                      ...                     ...   
7038                  No                      DSL                     Yes   
7039                 Yes              Fiber optic                      No   
7040                  No                      DSL                      No   
7041                  No                      DSL                     Yes   
7042    No phone service                      DSL                     Yes   

     internet.OnlineBackup internet.DeviceProtection internet.TechSupport  \
0                      Yes                        No                  Yes   
1                       No                        No                   No   
2                       No                       Yes                   No   
3                      Yes                       Yes                   No   
4                       No                        No                  Yes   
...                    ...                       ...                  ...   
7038                    No                        No                  Yes   
7039                    No                        No                   No   
7040                   Yes                        No                   No   
7041                    No                       Yes                  Yes   
7042                   Yes                       Yes                   No   

     internet.StreamingTV internet.StreamingMovies account.Contract  \
0                     Yes                       No         One year   
1                      No                      Yes   Month-to-month   
2                      No                       No   Month-to-month   
3                     Yes                      Yes   Month-to-month   
4                     Yes                       No   Month-to-month   
...                   ...                      .

In [37]:
df_aux.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Churn                      7032 non-null   int64  
 1   customer.gender            7032 non-null   int64  
 2   customer.SeniorCitizen     7032 non-null   int64  
 3   customer.Partner           7032 non-null   int64  
 4   customer.Dependents        7032 non-null   int64  
 5   customer.tenure            7032 non-null   int64  
 6   phone.PhoneService         7032 non-null   int64  
 7   phone.MultipleLines        7032 non-null   object 
 8   internet.InternetService   7032 non-null   object 
 9   internet.OnlineSecurity    7032 non-null   object 
 10  internet.OnlineBackup      7032 non-null   object 
 11  internet.DeviceProtection  7032 non-null   object 
 12  internet.TechSupport       7032 non-null   object 
 13  internet.StreamingTV       7032 non-null   object 
 1

In [38]:
df_dummies = pd.get_dummies(df_aux, dtype=np.int64).copy()
df_dummies

Churn  customer.gender  customer.SeniorCitizen  customer.Partner  \
0         0                1                       0                 1   
1         0                0                       0                 0   
2         1                0                       0                 0   
3         1                0                       1                 1   
4         1                1                       1                 1   
...     ...              ...                     ...               ...   
7038      0                1                       0                 0   
7039      1                0                       0                 1   
7040      0                0                       0                 0   
7041      0                0                       0                 1   
7042      0                0                       0                 1   

      customer.Dependents  customer.tenure  phone.PhoneService  \
0                       1                9                   1   
1                       0                9                   1   
2                       0                4                   1   
3                       0               13                   1   
4                       0                3                   1   
...                   ...              ...                 ...   
7038                    0               13                   1   
7039                    0               22                   1   
7040                    0                2                   1   
7041                    1               67                   1   
7042                    1               63                   0   

      account.PaperlessBilling  account.Charges.Monthly  \
0                            1                    65.60   
1                            0                    59.90   
2                            1                    73.90   
3                            1                    98.00   
4                            1                    83.90   
...                        ...                      ...   
7038                         0                    55.15   
7039                         1                    85.10   
7040                         1                    50.30   
7041                         0                    67.85   
7042                         0                    59.00   

      account.Charges.Total  ...  internet.StreamingMovies_No  \
0                    590.40  ...                            1   
1                    539.10  ...                            0   
2                    295.60  ...                            1   
3                   1274.00  ...                            0   
4                    251.70  ...                            1   
...                     ...  ...                          ...   
7038                 716.95  ...                            1   
7039                1872.20  ...                            0   
7040                 100.60  ...                            1   
7041                4545.95  ...                            0   
7042                3717.00  ...                            0   

      internet.StreamingMovies_No internet service  \
0                                                0   
1                                                0   
2                                                0   
3                                                0   
4                                                0   
...                                            ...   
7038                                             0   
7039                                             0   
7040                                             0   
7041                                             0   
7042                                             0   

      internet.StreamingMovies_Yes  account.Contract_Month-to-month  \
0                                0                                0   
1                                1         

In [39]:
df_dummies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 41 columns):
 #   Column                                           Non-Null Count  Dtype  
---  ------                                           --------------  -----  
 0   Churn                                            7032 non-null   int64  
 1   customer.gender                                  7032 non-null   int64  
 2   customer.SeniorCitizen                           7032 non-null   int64  
 3   customer.Partner                                 7032 non-null   int64  
 4   customer.Dependents                              7032 non-null   int64  
 5   customer.tenure                                  7032 non-null   int64  
 6   phone.PhoneService                               7032 non-null   int64  
 7   account.PaperlessBilling                         7032 non-null   int64  
 8   account.Charges.Monthly                          7032 non-null   float64
 9   account.Charges.Total              

### Traduzindo Colunas:

In [40]:
dicionario_nome_colunas = { 
'gender': 'gênero', 
'SeniorCitizen': 'idoso',
'Partner': 'tem_parceiro',
'Dependents': 'tem_dependentes',
'tenure': 'meses_de_contrato',
'PhoneService': 'servico_telefonico',
'MultipleLines': 'multiplas_linhas',
'InternetService': 'servico_de_internet',
'OnlineSecurity': 'seguranca_online',
'OnlineBackup': 'backup_online',
'DeviceProtection': 'protecao_de_dispositivo',
'TechSupport': 'suporte_tecnico',
'StreamingTV': 'TV_a_cabo',
'StreamingMovies': 'assinatura_streaming',
'Contract': 'tipo_de_contrato',
'PaperlessBilling': 'fatura_online',
'PaymentMethod': 'forma_de_pagamento',
'Charges.Monthly': 'valor_por_mes',
'Charges.Total': 'valor_total',
'No internet service': 'sem_servico_internet',
'No phone service': 'sem_servico_telefone',
'Fiber optic': 'fibra_optica',
'customer': 'cliente',
'phone': 'telefone',
'account': 'conta',
'No': 'nao',
'Yes': 'sim',
'Month-to-month': 'mes_a_mes',
'One year': 'um_ano',
'Two year': 'dois_anos',
'Bank transfer': 'transferencia_bancaria',
'Credit card': 'cartao_de_credito',
'Electronic check': 'cheque_eletronico',
'Mailed check': 'cheque_por_correio',
'automatic': 'automatico'
}

In [41]:
for coluna_original, traducao in dicionario_nome_colunas.items():
    df_dummies.columns = df_dummies.columns.str.replace(coluna_original, traducao)
df_dummies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 41 columns):
 #   Column                                                        Non-Null Count  Dtype  
---  ------                                                        --------------  -----  
 0   Churn                                                         7032 non-null   int64  
 1   cliente.gênero                                                7032 non-null   int64  
 2   cliente.idoso                                                 7032 non-null   int64  
 3   cliente.tem_parceiro                                          7032 non-null   int64  
 4   cliente.tem_dependentes                                       7032 non-null   int64  
 5   cliente.meses_de_contrato                                     7032 non-null   int64  
 6   telefone.servico_telefonico                                   7032 non-null   int64  
 7   conta.fatura_online                                           7032 non-nul

### Inserindo Nova Coluna:

In [42]:
df_dummies.insert(8, 'conta.valor_por_dia', '')
df_dummies

Churn  cliente.gênero  cliente.idoso  cliente.tem_parceiro  \
0         0               1              0                     1   
1         0               0              0                     0   
2         1               0              0                     0   
3         1               0              1                     1   
4         1               1              1                     1   
...     ...             ...            ...                   ...   
7038      0               1              0                     0   
7039      1               0              0                     1   
7040      0               0              0                     0   
7041      0               0              0                     1   
7042      0               0              0                     1   

      cliente.tem_dependentes  cliente.meses_de_contrato  \
0                           1                          9   
1                           0                          9   
2                           0                          4   
3                           0                         13   
4                           0                          3   
...                       ...                        ...   
7038                        0                         13   
7039                        0                         22   
7040                        0                          2   
7041                        1                         67   
7042                        1                         63   

      telefone.servico_telefonico  conta.fatura_online conta.valor_por_dia  \
0                               1                    1                       
1                               1                    0                       
2                               1                    1                       
3                               1                    1                       
4                               1                    1                       
...                           ...                  ...                 ...   
7038                            1                    0                       
7039                            1                    1                       
7040                            1                    1                       
7041                            1                    0                       
7042                            0                    0                       

      conta.valor_por_mes  ...  internet.assinatura_streaming_nao  \
0                   65.60  ...                                  1   
1                   59.90  ...                                  0   
2                   73.90  ...                                  1   
3                   98.00  ...                                  0   
4                   83.90  ...                                  1   
...                   ...  ...                                ...   
7038                55.15  ...                                  1   
7039                85.10  ...                                  0   
7040                50.30  ...                                  1   
7041                67.85  ...                                  0   
7042                59.00  ...                                  0   

      internet.assinatura_streaming_sem_servico_internet  \
0                                                     0    
1                                                     0    
2                                                     0    
3                                                     0    
4                                                     0    
...                                                 ...    
7038                                                  0    
7039                                                  0    
7040                                                  0    
7041                                                  0    
7042                                      

In [43]:
df_dummies['conta.valor_por_dia'] = (df_dummies['conta.valor_por_mes']*12/365).round(2)
df_dummies

Churn  cliente.gênero  cliente.idoso  cliente.tem_parceiro  \
0         0               1              0                     1   
1         0               0              0                     0   
2         1               0              0                     0   
3         1               0              1                     1   
4         1               1              1                     1   
...     ...             ...            ...                   ...   
7038      0               1              0                     0   
7039      1               0              0                     1   
7040      0               0              0                     0   
7041      0               0              0                     1   
7042      0               0              0                     1   

      cliente.tem_dependentes  cliente.meses_de_contrato  \
0                           1                          9   
1                           0                          9   
2                           0                          4   
3                           0                         13   
4                           0                          3   
...                       ...                        ...   
7038                        0                         13   
7039                        0                         22   
7040                        0                          2   
7041                        1                         67   
7042                        1                         63   

      telefone.servico_telefonico  conta.fatura_online  conta.valor_por_dia  \
0                               1                    1                 2.16   
1                               1                    0                 1.97   
2                               1                    1                 2.43   
3                               1                    1                 3.22   
4                               1                    1                 2.76   
...                           ...                  ...                  ...   
7038                            1                    0                 1.81   
7039                            1                    1                 2.80   
7040                            1                    1                 1.65   
7041                            1                    0                 2.23   
7042                            0                    0                 1.94   

      conta.valor_por_mes  ...  internet.assinatura_streaming_nao  \
0                   65.60  ...                                  1   
1                   59.90  ...                                  0   
2                   73.90  ...                                  1   
3                   98.00  ...                                  0   
4                   83.90  ...                                  1   
...                   ...  ...                                ...   
7038                55.15  ...                                  1   
7039                85.10  ...                                  0   
7040                50.30  ...                                  1   
7041                67.85  ...                                  0   
7042                59.00  ...                                  0   

      internet.assinatura_streaming_sem_servico_internet  \
0                                                     0    
1                                                     0    
2                                                     0    
3                                                     0    
4                                                     0    
...                                                 ...    
7038                                                  0    
7039                                                  0    
7040                                                  0    
7041                                                  0    
7042                          